### PR TITLE
upstream: support max connection duration for upstream HTTP connections

### LINKED ISSUE #15107: Upstream max connection duration should be configurable.
We currently support max connection duration for downstream connections, but not yet for upstream connections. Allowing upstream connections to have a max duration provides users a guarantee that after a certain amount of time, connections will be cycled. This allows for service discovery through Kubernetes Service IPs, for example, to work more reliably.

### PR DESCRIPTION
**Pull request by esmet**
Commit Message: upstream: support max connection duration for upstream connections
Risk Level: low, isolated opt-in feature
Testing: new unit tests
Docs Changes: updated protobuf definition to no longer mention that max lifetime is not implemented for upstream connections
Release Notes: upstream: support max connection duration for upstream connections
Platform Specific Features:
Fixes #15107

### CODE REVIEW

**Review by htuch — COMMENTED** [review]
/lgtm api

File: test/common/upstream/upstream_impl_test.cc
Comment (jmarantz) [inline test/common/upstream/upstream_impl_test.cc:2817]: I was expecting to see a test using simulated time, where we advance time forward past the timeout and then verify that the connections get properly closed.

**Review by jmarantz — COMMENTED** [review]
looks good but would like a test that pulls everything together with SImulatedTime.

File: test/common/upstream/upstream_impl_test.cc
Comment (esmet) [reply test/common/upstream/upstream_impl_test.cc:2817]: I updated my new tests to start using a derived ConnPoolImplDispatcherBaseTest that uses a real (non-mock) dispatcher and simulated time. I left the existing mock dispatcher to avoid modifying existing tests, especially the ones that derive from this base test.

**Comment by esmet** [comment]
I'm working on a fix for the ASAN failure in the new test.

File: source/common/conn_pool/conn_pool_base.cc
Comment (jmarantz) [inline source/common/conn_pool/conn_pool_base.cc:477]: can you use an explicit type here?

File: source/common/conn_pool/conn_pool_base.cc
Comment (jmarantz) [inline source/common/conn_pool/conn_pool_base.cc:478]: `!= 0`

File: envoy/upstream/upstream.h
Comment (jmarantz) [inline envoy/upstream/upstream.h:556]: `max_duration_exceeded` ?

Should we have doc for this new stat?

File: test/common/conn_pool/conn_pool_base_test.cc
Comment (jmarantz) [inline test/common/conn_pool/conn_pool_base_test.cc:114]: any reason not to dynamic_cast & ASSERT_TRUE here?

File: test/common/conn_pool/conn_pool_base_test.cc
Comment (jmarantz) [inline test/common/conn_pool/conn_pool_base_test.cc:113]: nit: s/-> void//

File: test/common/conn_pool/conn_pool_base_test.cc
Comment (jmarantz) [inline test/common/conn_pool/conn_pool_base_test.cc:289]: nit: EXPECT_TRUE

File: test/common/conn_pool/conn_pool_base_test.cc
Comment (jmarantz) [inline test/common/conn_pool/conn_pool_base_test.cc:328]: check that the stat you added gets bumped?

File: test/common/upstream/upstream_impl_test.cc
Comment (jmarantz) [reply test/common/upstream/upstream_impl_test.cc:2817]: thanks!

File: test/common/conn_pool/conn_pool_base_test.cc
Comment (jmarantz) [inline test/common/conn_pool/conn_pool_base_test.cc:338]: these tests are long and fairly similar. WDYT of factoring out some helper functions that you could put in the tset-class that the tests can share, and would differ only by a few params or which helper methods need to be called?

That might make the differences between the tests easier to spot when reading the code.

**Review by jmarantz — COMMENTED** [review]
This basically looks great; just left a few minor comments.

I do think the senior maintainer should be someone pretty familiar with this area of the data plane code but I'll do one more round before passing it off.

**Comment by jmarantz** [comment]
also needs main merge.

/wait

File: test/common/conn_pool/conn_pool_base_test.cc
Comment (esmet) [reply test/common/conn_pool/conn_pool_base_test.cc]: I went ahead and refactored out a lot of the boilerplate so now the test machinery is better separated from the test cases. The result is that tests are shorter and easier to understand.

File: test/common/conn_pool/conn_pool_base_test.cc
Comment (jmarantz) [reply test/common/conn_pool/conn_pool_base_test.cc:114]: maybe just factor this out until a new function `incrementActiveStreams(&client)` and you can do the dynamic-cast and ASSERT in there?

File: test/common/upstream/upstream_impl_test.cc
Comment (jmarantz) [inline test/common/upstream/upstream_impl_test.cc:2847]: are these two strings identical except the "0s" vs "9s"? You could use fmt::format to factor that out too :)

**Review by jmarantz — COMMENTED** [review]
This is a lot nicer now! Ping when the other comments are resolved and I"ll do another full pass.

Thanks!

/wait

File: source/common/conn_pool/conn_pool_base.cc
Comment (esmet) [reply source/common/conn_pool/conn_pool_base.cc:477]: Added the full type (an optional) and updated the conditional to check has_value()

File: test/common/conn_pool/conn_pool_base_test.cc
Comment (esmet) [reply test/common/conn_pool/conn_pool_base_test.cc:334]: That check is at L327 unless you had another stat in mind

File: test/common/upstream/upstream_impl_test.cc
Comment (esmet) [reply test/common/upstream/upstream_impl_test.cc:2847]: Done!

**Comment by esmet** [comment]
@jmarantz this should now be ready for another pass

**Review by jmarantz — APPROVED** [review]
looks great; just a few nits. I think we need a data-plane expert as the senior maintainer.

I'm casting for the right one out-of-band and will assign once we find one.

File: test/common/conn_pool/conn_pool_base_test.cc
Comment (esmet) [reply test/common/conn_pool/conn_pool_base_test.cc:188]: These are technically mutable so that individual tests can change them as appropriate for the test.

File: test/common/conn_pool/conn_pool_base_test.cc
Comment (esmet) [reply test/common/conn_pool/conn_pool_base_test.cc:190]: same re: mutable for tests. In particular, we change this to `1` for a few tests to force a conn pool to enter the draining state immediately after creating a single ActiveClient

File: source/common/conn_pool/conn_pool_base.cc
Comment (esmet) [reply source/common/conn_pool/conn_pool_base.cc:614]: I looked at this a little harder. If this were to happen in real life, it would be similar to reaching the connect timeout, and the connection would `close()`. However, I think there might be trouble since we're moving the ActiveClient to the busy_clients_ list (not the connecting_clients_ list) and in theory the connect timer could still fire and try to `close()` again. Not entirely sure if either those are problems in practice but it definitely doesn't feel right.

So I think the best two options are to upgrade this to a fatal RELEASE_ASSERT or to keep it as an ASSERT(presumably only fatal in dev builds and tests) and guarding against the bad state by doing nothing if the state is, in fact, CONNECTING. I'm leaning towards the latter because there's no real harm if we hit this state, outside the fact that it means the original feature isn't working correctly :)

File: test/common/conn_pool/conn_pool_base_test.cc
Comment (esmet) [reply test/common/conn_pool/conn_pool_base_test.cc:78]: Yes there was another instance in conn_pool_base_test.cc and I fixed that up, too.

File: source/common/conn_pool/conn_pool_base.cc
Comment (jmarantz) [reply source/common/conn_pool/conn_pool_base.cc:614]: I'm OK with the latter; can we hit that in a testcase somehow?

File: test/common/conn_pool/conn_pool_base_test.cc
Comment (jmarantz) [reply test/common/conn_pool/conn_pool_base_test.cc:188]: sounds good; sorry for the noise :)

**Review by jmarantz — APPROVED** [review]
lgtm modulo still-unresolved comments.

/assign-from @envoyproxy/senior-maintainers

File: source/common/conn_pool/conn_pool_base.cc
Comment (esmet) [reply source/common/conn_pool/conn_pool_base.cc:614]: According to STYLE.md, this case is probably best covered as an ENVOY_BUG and we can test it using EXPECT_ENVOY_BUG.

File: docs/root/version_history/current.rst
Comment (alyssawilk) [inline docs/root/version_history/current.rst:135]: this should probably reflink the field

Comment (esmet) [reply docs/root/version_history/current.rst]: I added a reflink and I also fixed it to mention "max connection duration" not "max stream duration"

**Comment by esmet** [comment]
@jmarantz @alyssawilk comments addressed and there's a test for the new ENVOY_BUG (formerly ASSERT)

**Comment by jmarantz** [comment]
I think if you 'merge main' a common flake in 'coverage' will be reserved.

**Comment by esmet** [comment]
!retest

File: api/envoy/config/core/v3/protocol.proto
Comment (alyssawilk) [inline api/envoy/config/core/v3/protocol.proto:160]: why the change?  AFIK the drain sequence kicks in for HTTP/1 too, which is sending connection:close on responses and force closing.

File: source/common/conn_pool/conn_pool_base.cc
Comment (alyssawilk) [inline source/common/conn_pool/conn_pool_base.cc:612]: onConnectionDurationTimeout would be more consistent.  ditto on the timer and the logs.

File: source/common/conn_pool/conn_pool_base.cc
Comment (alyssawilk) [inline source/common/conn_pool/conn_pool_base.cc:617]: Very nice.  I had a long comment about how there was a lifetime bug because I missed this, until I saw that you tested it already and went back to figure out why it wasn't a problem :-)

optionally, I would suggest disarming the timer in onConnectionEvent where we disable the connection timer.  It feels safer to not have an alarm fire once the session is on the deferred delete list, WDYT?

**Review by alyssawilk — COMMENTED** [review]
Very nice.  Along with the comments below I'd suggest a test addition to idle_timeouts_integration_test.  I think this is solid but I find having tests of data plane lifetime interactions is generally worthwhile :-)
/wait

File: api/envoy/config/core/v3/protocol.proto
Comment (esmet) [reply api/envoy/config/core/v3/protocol.proto:160]: I was reading the docs and code for idle_timeout and felt this matched. It was also important to call out that the drain sequence won't kick in for upstream connections, only downstream ones.

File: source/common/conn_pool/conn_pool_base.cc
Comment (esmet) [reply source/common/conn_pool/conn_pool_base.cc:617]: Yes it would definitely be safest to explicitly disable the timer in that case. I'll take a look at that.

File: source/common/conn_pool/conn_pool_base.cc
Comment (esmet) [reply source/common/conn_pool/conn_pool_base.cc:612]: I originally wanted to be consistent with the max streams variable name (which refers to  "lifetime") but it probably makes more sense to be consistent with the API field names.

File: api/envoy/config/core/v3/protocol.proto
Comment (alyssawilk) [reply api/envoy/config/core/v3/protocol.proto:160]: Why don't you think it applies to HTTP/1 and HTTP/3 though?

Comment (esmet) [reply api/envoy/config/core/v3/protocol.proto:160]: I looked at the implementation and noted a call to `goAway()` which likely includes H3 but it made me think that it didn't apply to H1 (but, mostly because of the existing docs on idle_timeout).

The addition of "HTTP/2 downstream connections" instead of just "downstream connections" is possibly an overreach. The main point from my change is that upstream connections won't be subject to a drain_timeout.

Comment (esmet) [reply api/envoy/config/core/v3/protocol.proto:160]: For example, onConnectionDurationTimeout() and onIdleTimeout both startDrainSequence() and call codec_->shutdownNotice() (appears to no-op for H1) and later, on drain timeout, codec_->goAway() (also appears to no-op for H1).

But now that I read this more, I see that the subsequent call to checkForDeferredClose() is likely responsible for sending Connection: close downstream.

So now I'm curious why the idle_timeout docs are worded the way they are. Maybe there right thing to do is simply remove the "HTTP/2" distinction from idle_timeout since it appears that draining happens for all. Thoughts?

Comment (alyssawilk) [reply api/envoy/config/core/v3/protocol.proto:160]: yeah I'd be inclined to remove the HTTP/2
again while the shutdownNotice doesn't do anything for HTTP/1.1 the transition to draining results in connection: close which is the http/1.1 style drain, so I think the sequence of drain followed by force-close works for http 1.1, 2, and 3, just in different ways.

**Review by alyssawilk — APPROVED** [review]
Nice!

**Comment by alyssawilk** [comment]
Unfortunately looks like this needs a main merge which should hopefully fix both the conflict and the coverage flake.

**Comment by esmet** [comment]
> Unfortunately looks like this needs a main merge which should hopefully fix both the conflict and the coverage flake.

Yep and I still need to address your comment about adding an integration test

**Pull request merged**
